### DISCUSSION TITLE
code generation frequently

### DISCUSSION THREAD

**Discussion by kannae97** [TabbyML/tabby#2938] [Ideas]
In the code generation scenario, there are often some invalid generation, such as the deletion of some indentation characters. I am not sure if this is what you expected. If this is an optimization point, does the tabby official team have plans to optimize this scenario?

---
Please reply with a 👍 if you want this feature.

**Comment by wsxiaoys** [comment]
Hi - please share your tabby setup: e.g what kind of model are you using for code completion.

Besides, can you provide actual cases that you feel could use improvements? e.g in form of screenshort / recording.

**Comment by kannae97** [comment]
> Hi - please share your tabby setup: e.g what kind of model are you using for code completion.
> 
> Besides, can you provide actual cases that you feel could use improvements? e.g in form of screenshort / recording.

Hi，thanks for your answer, I use code-llama-7b, and the context uses 40 lines. I have currently modified the code on the plugin side and reduced some calls that I think are unnecessary (not necessarily correct, but I think it is more reasonable). The rules are as follows:
1. Each time the completion interface is called, the plugin side records the generated completion content and line number (only the latest completion result is recorded).
2. When there are operations such as deleting indentation characters, adding indentation characters, adding spaces, etc., determine whether the line where the cursor is located is an empty string (this rule is mainly to solve the problem of repeated calls in new lines). If this rule is hit, find out from the latest cached results whether the current line already has cached content. If so, display the recommended content fixedly.

I tried to modify the code in this way, but unfortunately, the content displayed every time I entered would shake, the actual number of calls dropped significantly, and indicators such as the adoption rate also increased significantly (I think it is closer to real user feedback and can better reflect usability😂).

**Reply by wsxiaoys** [reply]
Hi - it's a bit challenging to understand your idea. If you have time, would you please share some screenshots or a recording regarding the issue you want to solve? For example, in what cases is ghost text displayed when you really don't want it to be?

**Comment by 5bug** [comment]
Compared to mainstream products, Tabby's code completion is indeed more frequent.
I hope recommendations can be triggered only when necessary.
